File: RPNCalculator3.java
// 逆ポーランド記法電卓 (関数型スタイル2)

import java.util.Arrays;
import java.util.ArrayDeque;
import java.util.Collections;
import java.util.Deque;
import java.util.HashMap;
import java.util.Map;
import java.util.function.BiFunction;
import java.util.function.BinaryOperator;

class RPNCalculator3 {
  private static final Map<String, BinaryOperator<Double>> OPERATORS;

  static {
    final Map<String, BinaryOperator<Double>> map = new HashMap<>();
    map.put("+", (x, y) -> y + x);
    map.put("-", (x, y) -> y - x);
    map.put("*", (x, y) -> y * x);
    map.put("/", (x, y) -> y / x);
    OPERATORS = Collections.unmodifiableMap(map);
  }

  public static double calculate(final String rpn) {
    final BiFunction<Deque<Double>, String, Deque<Double>> calc = (stack, x) -> {
      if (x.chars().allMatch(Character::isDigit)) {
        stack.addFirst(Double.parseDouble(x));
        return stack;
      } else if (stack.size() >= 2) {
        final double res = opeFn(x).apply(stack.removeFirst(), stack.removeFirst());
        stack.addFirst(res);
        return stack;
      } else {
        throw new IllegalArgumentException("unexpected pattern found");
      }
    };
    return Arrays.stream(rpn.split("\\s+"))
      .reduce(new ArrayDeque<>(), calc, (l, r) -> {
        l.addAll(r);
        return l;
      })
      .removeFirst();
  }

  private static BinaryOperator<Double> opeFn(final String ope) {
    if (OPERATORS.containsKey(ope)) {
      return OPERATORS.get(ope);
    } else {
      throw new IllegalArgumentException(
        String.format("unsupported operator '%s' is used", ope));
    }
  }

  public static void main(final String[] args) {
    // 利用例
    final String rpn = "1 2 + 3 / 4 - 5 *";
    System.out.printf("%s%n = %s%n", rpn, calculate(rpn));
  }
}

File: rpn_calculator3.py
# 逆ポーランド記法電卓 (関数型スタイル2)

from functools import reduce
from operator import add, sub, mul, truediv

_OPERATORS = {
    "+": add,
    "-": sub,
    "*": mul,
    "/": truediv
}


def calculate(rpn):
    def calc(stack, x):
        if x.isdigit():
            return [int(x)] + stack
        elif len(stack) >= 2:
            return [_ope_fn(x)(stack[1], stack[0])] + stack[2:]
        else:
            raise ValueError("unexpected pattern found")
    return reduce(calc, rpn.split(), [])[0]


def _ope_fn(ope):
    if ope in _OPERATORS:
        return _OPERATORS[ope]
    else:
        raise ValueError("unsupported operator '{0}' is used".format(ope))


if __name__ == "__main__":
    # 利用例
    rpn = "1 2 + 3 / 4 - 5 *"
    print(rpn, "\n = ", calculate(rpn))
